### PR TITLE
Add context managers for handling VTK errors

### PR DESCRIPTION
**Pull request by whophil**
### Overview
We need a way to catch VTK errors, both in tests and production code. Inspired by discussion in #1374

It is possible to catch VTK errors with `Observer`, but:
1. it can be challenging to attach the Observer to a desired algorithm instance from the test side.
2. the Observer sometimes declares that no error has been raised, even though one has (see https://github.com/pyvista/pyvista/pull/1374#discussion_r650448650)

 

### Details

This PR adds two context managers for handling VTK errors. These context managers redirect VTK output to a temporary file. Upon exit, they set the VTK output to whatever it was before, and clean up the temporary file.

The basic usage is as follows:
```python
# The VTK error appears on stderr
# A RuntimeError is raised
with pyvista.utilities.errors.raise_vtk_errors():
    generate_vtk_err()

# The VTK error appears on stderr
# No error is raised, but the output is logged
error_catcher = pyvista.utilities.errors.VtkErrorCatcher(raise_error=False)
with error_catcher:
    # this function generates a VTK error
    generate_vtk_err()
print('logged error is\n', error_catcher.log)
```

`raise_vtk_errors` will raise a RuntimeError for any VTK error raised within its context. This is a shorthand method using `VtkErrorCatcher` under the hood. If you'd like to parse VTK errors, you will need to use the `VtkErrorCatcher` directly, which makes the log string available outside of the context.

### CODE REVIEW

**Comment by MatthewFlamm** [comment]
Nice! I would vote for the context manager to be more exposed to users, e.g. by importing in `pyvista.utilities`?  If so, you should add an example into the docstring.

File: tests/test_utilities.py
Comment (MatthewFlamm) [inline tests/test_utilities.py:491]: This example will not throw a vtk error if #1374 is merged.

Comment (MatthewFlamm) [reply tests/test_utilities.py:491]: Oops nevermind.  This will certainly work.

**Comment by whophil** [comment]
In tracking down some test abnormalities, I found that global VTK error handling was being altered in two other places. Given that we are now trying to catch VTK errors, I think we'll need to get rid of anything which modifies global VTK error handling. I removed these lines in https://github.com/pyvista/pyvista/pull/1382/commits/3d9c637571f2512dce8277eab1897f9211ad7290.

I believe that error handling was turned off there for aesthetic reasons, likely to avoid cluttering stdout/stderr. @rodrigomologni can you please confirm that there is no other functional reason for disabling the "global warning display"?

**Comment by akaszynski** [comment]
> Given that we are now trying to catch VTK errors, I think we'll need to get rid of anything which modifies global VTK error handling. I removed these lines in 3d9c637.

Agreed with this, and I really like having uniform global error handling.

Merged `master` into this branch.  This will help with CI, and thanks for your earlier PR.

File: pyvista/utilities/errors.py
Comment (akaszynski) [inline pyvista/utilities/errors.py:31]:
```suggestion
    """Catch VTK errors, using a context manager to temporarily redirect errors to a tempfile.
    
    Parameters
    ----------
    raise_errors : bool, optional
        Raise a ``RuntimeError`` when a VTK error is encountered.  Defaults to ``False``.

    Examples
    --------
    Catch VTK errors using the context manager.

    >>> import pyvista
    >>> error_catcher = pyvista.utilities.errors.VtkErrorCatcher()
    >>> with error_catcher:
    ...     sphere = pyvista.Sphere()
    """
    
```

File: pyvista/utilities/errors.py
Comment (akaszynski) [inline pyvista/utilities/errors.py:52]:
```suggestion
```

Python methods return ``None`` by default, so we don't have to add this here.

File: pyvista/utilities/errors.py
Comment (akaszynski) [inline pyvista/utilities/errors.py:65]:
```suggestion
```

**Review by akaszynski — APPROVED** [review]
I'm happy with this, but we might consider adding logging using the `logging` module and always reading the log file and outputting it to the `logging` module at the ERROR level.

**Comment by rodrigomologni** [comment]
> I believe that error display was turned off there for aesthetic reasons, likely to avoid cluttering stdout/stderr. @rodrigomologni can you please confirm that there is no other functional reason for disabling the "global warning display"?

Hi @akaszynski! There is no other reason. It's just aesthetic.

**Comment by whophil** [comment]
> I'm happy with this, but we might consider adding logging using the `logging` module and always reading the log file and outputting it to the `logging` module at the ERROR level.

Thanks for the review! Your comment inspired me to do this a different way, using `Observer` to watch the a `vtkStringOutputWindow()`, as was already done in `send_errors_to_logging()`. This supports Python's logging because `Observer` does. It also makes all the tempfile stuff unnecessary.

To suppor this I made `Observer` optionally keep track of history. Lastly, if VTK errors are raised within the context, `RuntimeError` returns a list of the errors rather than a single `str`-joined error, in case somebody wants to do more sophisticated error handling down the line.

**Comment by whophil** [comment]
@MatthewFlamm these context managers are now exposed in `pyvista.utilities`. Re-requesting review - I'm ready to merge if you guys are, so feel free to hit the green button once you've approved.

File: pyvista/utilities/errors.py
Comment (MatthewFlamm) [inline pyvista/utilities/errors.py:45]:
```suggestion
    >>> with pyvista.VtkErrorCatcher() as error_catcher:
```

File: pyvista/utilities/errors.py
Comment (MatthewFlamm) [inline pyvista/utilities/errors.py:83]:
```suggestion
    >>> with pyvista.raise_vtk_errors():
```

File: pyvista/utilities/errors.py
Comment (MatthewFlamm) [inline pyvista/utilities/errors.py:76]: I think it may be confusing to users to have more than one context manager option.  It took me a careful read of the parameters and descriptions to get it.  I think either a short guide in one or both docstrings on what the difference is, or collapsing into a single context manager, would be helpful.

**Review by MatthewFlamm — APPROVED** [review]
Looks good.  

All of my comments are suggestions.  I will not merge as I'm not 100% familiar with the Observer functionality and you have changed it since the last approval.

File: pyvista/utilities/errors.py
Comment (akaszynski) [reply pyvista/utilities/errors.py:75]: I'd also prefer having a single context manager.  @whophil, do you think this is possible?

Comment (whophil) [reply pyvista/utilities/errors.py:75]: `raise_vtk_errors` is just a convenience method which wraps  `VtkErrorCatcher`. However, you can use `VtkErrorCatcher` directly to raise an error if you'd like, it's just slightly more wordy.

```python
# this will make errors available after the context in `vtk_errors`
vtk_errors = pyvista.VtkErrorCatcher()
with vtk_errors:
    generate_vtk_error()

# this will just raise the errors
with pyvista.VtkErrorCatcher(raise_errors=True):
    generate_vtk_error()

# this is the shorthand way, which we could consider removing
with pyvista.raise_vtk_errors():
    generate_vtk_error()
```

If you guys prefer, we can either:
1. Remove `raise_vtk_error` entirely
2. Remove `raise_vtk_error` from the "public"-interface by not importing it in `pyvista.utilities`

Comment (MatthewFlamm) [reply pyvista/utilities/errors.py:75]: I vote for number 1.

Comment (akaszynski) [reply pyvista/utilities/errors.py]: Agreed with @MatthewFlamm.  From `import this`:

> There should be one-- and preferably only one --obvious way to do it.

**Review by akaszynski — APPROVED** [review]
Good work.  Thanks!

I've added a section in our docs explaining the class.  These changes update the entire utilities API docs and was *slightly* out of scope of this PR, but I didn't want these changes to go unnoticed to the end-user without docs.

sc

**Comment by whophil** [comment]
Thanks for the help @akaszynski and @MatthewFlamm ! Everything looks good to me, please merge when you're ready

**Comment by tkoyama010** [comment]
I am +100 for this. Thanks for PR 😄

**Pull request merged**
